File: charge.py
#!/usr/bin/env python2.7
import numpy as np
import h5py
import ROOT
import time

''' Code to analyze .hdf5 files 
    that store the fully digitized
    scope waveform 
'''
# TO DO: improve speed

canvas = None
termination_ohms = 50 

# Lognormal distribution for the pulse fit 
class lognormal:
    def __call__(self,x,par):
        y = np.exp(-0.5 * (np.log(x[0] / par[0]) / par[1])**2.0)
        y *= par[2] / (x[0] * par[1] * np.sqrt(2.0 * np.pi))
        p = np.exp(-0.5 * (np.log(x[0] / par[3]) / par[4])**2.0)
        p *= par[5] / (x[0] * par[4] * np.sqrt(2.0 * np.pi))
        q = y + p
        return -q

'''  Get the waveform attributes
   : param filename:   data file
   : param channel:    data channel
     Returns number of samples, number of waveforms
     horizontal, and verical resolutions
'''
def read_hdf5(filename, channel):
    samples = len(channel[0])
    waveforms = len(channel)
    dy = channel.attrs['vertical_gain']
    dx = channel.attrs['horiz_interval']
    return samples, waveforms, dy, dx

'''  Find the bin corresponding to the minimum voltage 
     Also returns the pedestal for the given waveform 
'''
def peak_finder(i, samples, channel, dy, pedestal_window):
    min_voltage = 0.0
    pedestal = np.mean(channel[i,:pedestal_window])*dy
    for j in range(0, samples):
        voltage = channel[i,j]*dy - pedestal
        if voltage < min_voltage:
            min_voltage = voltage
            min_bin = j

    return min_bin, pedestal

''' Returns an array with all the channels
    Ex: ['channel1','channel2',channel3']
'''
def channels(num_channels):
    channel = []
    for i in range(1,num_channels+1):
        channel.append("channel"+str(i))

    return channel

'''  Draw waveforms with the debug option 
   : param i:            waveform number 
   : param samples:      number of samples
   : param channel:      channel
   : param dy:           veritcal resolution
   : param min_bin:      bin correspong to peak
   : param pedestal:     length of pedestal window in time bins
   : param num_channels: number of channels to draw (max 2)
   : param fit:          fit the trigger waveform w/ lognormal
     Returns TH1 Waveform of scope trace
'''

def draw_waveform(f, i, samples, channel, dy, dx, min_bin, pedestal, num_channels, fit):
    global canvas
    canvas = ROOT.TCanvas()
    pad1 = ROOT.TPad("pad1","pad1",0,0.5,1,1)
    pad2 = ROOT.TPad("pad2","pad2",0,0.5,1,0.0)
    Waveform = ROOT.TH1F('waveform',';t(ns);Voltage',samples,0,samples*0.1)
    channel_list = channels(num_channels)
    for ch in channel_list:
        c = f[ch]
        for j in range(0, samples):
            voltage = c[i,j]*dy - pedestal
            Waveform.SetBinContent(j+1,voltage)
            Waveform.SetBinError(j+1,dy)
        print "Drawing waveform",
        print i
        if ch == "channel1":
            print ch
            yaxis = ("Voltage (V) "+ch)
            pad1.SetBottomMargin(0)
            pad1.Draw()
            pad1.cd()
            Waveform.Draw("e")
            Waveform.SetStats(0)
            Waveform.SetLineColor(1)
            Waveform.GetYaxis().SetTitleOffset(1.20)
            Waveform.GetYaxis().SetTitle(yaxis)
            # do the fit for channel 1, and draw
            if fit==True:
                tmin, bottom_range, top_range = min_bin*dx*1e9, min_bin*dx*1e9-3.0, min_bin*dx*1e9+4.0
                fit = ROOT.TF1("lognormal",lognormal(),bottom_range,top_range,6)
                fit.SetParameter(0, tmin)
                fit.SetParameter(1, 0.02)
                fit.SetParameter(2, 0.1)
                fit.SetParameter(3, tmin)
                fit.SetParameter(4, 0.02)
                fit.SetParameter(5, 0.1)
                fit.SetParLimits(0, tmin-10, tmin+10)
                Waveform.Fit(fit,"0")
                fit.SetLineColor(ROOT.kSpring)
                fit.Draw("same")
                Waveform.GetXaxis().SetRangeUser(bottom_range-10,top_range+10)
            canvas.Update()
        else:
            print ch
            yaxis = ("Voltage (V) "+ch)
            canvas.cd()
            pad2.SetTopMargin(0)
            pad2.SetBottomMargin(0.1)
            pad2.Draw()
            pad2.cd()
            Waveform.Draw("e same")
            Waveform.SetStats(0)
            Waveform.SetLineColor(2)
            Waveform.GetYaxis().SetTitleOffset(1.20)
            Waveform.GetYaxis().SetTitle(yaxis)
            Waveform.GetXaxis().SetRangeUser(0,500)
            canvas.Update()
    #canvas.Update()
    raw_input()
    pad1.Close()
    pad2.Close()

''' In debug mode this calls draw_waveform, which will continue to
    print waveforms to the screen. Cancel with CTRL-C.
    In normal mode, this builds charge and pedestal histograms
    and write thems to a file.
'''
def plot(filename, channel, pedestal_window, num_channels, debug=True, fit=False):
    Tf = ROOT.TFile("output.root", 'recreate')
    Charge = ROOT.TH1F('charge',';Charge(pC);Counts', 1250,-1,10)
    Pedestal = ROOT.TH1F('pedestal',';Voltage(V);Counts', 50000,-1,1)

    f = h5py.File(filename)
    channel = f[channel]
    print "Number of channels", 
    print num_channels
    samples, waveforms, dy, dx = read_hdf5(f, channel)
        
    for i in range(0,waveforms):
        charge = 0
        min_bin, pedestal = peak_finder(i, samples, channel, dy, pedestal_window)
        if debug:
            draw_waveform(f, i, samples, channel, dy, dx, min_bin, pedestal, num_channels, fit)
        for j in range(min_bin-250,min_bin+250):
            voltage = channel[i,j]*dy - pedestal
            charge += (voltage*((-1000*dx*1e9)/termination_ohms))
        Charge.Fill(charge)
        Pedestal.Fill(pedestal)

    Tf.cd()
    Charge.Write()
    Pedestal.Write()
    Tf.Close()
    return 0

''' Input arguments to process
   : arg filename:        data file to process
   : arg channel:         channel to process
   : arg pedestal_window: length of the pedestal window, in time-bins
   : arg --debug:         write waveforms to screen
   : arg --num_channels:  number of channels to draw (currently only works for 2)
   : arg --fit:           fit the trigger signal with a lognormal
'''
if (__name__=="__main__"):
    import argparse

    parser = argparse.ArgumentParser(description='Draw waveforms')
    parser.add_argument('filename', help='Name of HDF5 file to process')
    parser.add_argument('channel', help='Name of channel to process')
    parser.add_argument('pedestal_window', type=int, help='Length of pedestal_window')
    parser.add_argument('--debug', action='store_true', help='Show debugging plots')
    parser.add_argument('--num_channels', type=int, default=1,
                         help='Number of channels to process')
    parser.add_argument('--fit', type=bool, default=False)
    args = parser.parse_args()
    plot(args.filename, args.channel, args.pedestal_window, args.num_channels, args.debug, args.fit)

File: charge.cpp
/* Build charge, pedestal, and average waveform histograms
   from oscilloscope data in .hdf5 format
   * * *
   * input params:
   * argv[1]: text file listing .hdf5 for analysis
   * argv[2]: name of output root file to write histograms to
   * argv[3]: length of the pedestal window
   * argv[4]: length of the integration window
   * argv[5]: scope channel for analysis
   * * *
*/

#include <iostream>
#include <string>
#include <fstream>
#include <math.h>
#include <vector>

// Root Libraries
#include <TROOT.h>
#include <TFile.h>
#include <TMath.h>
#include <TCanvas.h>
#include <TH1.h>

// HDF5 Library
#include "H5Cpp.h"

#ifndef H5_NO_NAMESPACE
using namespace H5;
#endif

using namespace std;

const int RANK_OUT = 2; // Data Rank

struct DataCluster{
    DataSet *dataset; // Dataset pointer
    DataSpace dataspace; // DataSet's DataSpace
    DataSpace memspace; // MemSpace Object for Data Extraction
    hsize_t offset[RANK_OUT]; // Data Extraction Parameters...
    hsize_t count[RANK_OUT];
    hsize_t offset_out[RANK_OUT];
    hsize_t count_out[RANK_OUT];
    unsigned long trace_length; // Length of a Scope Trace
    unsigned long n_traces; // Number of traces in DataSet
    char * data_out; // Pointer to Data Buffer
};

typedef struct DataCluster DataCluster;

// Initialize Datacluster
DataCluster * Init_Data(DataSet *dataset);

// For each trace, read in the dataset
int Read_Trace(DataCluster *datacluster, unsigned long trace_index);

// Return the voltage of the sample
double getVoltage(DataCluster *datacluster, double pedestal, double dy, int i);

// Return the integrated charge over the signal window
double getCharge(float pedestal_window, float signal_window, DataCluster *datacluster, double pedestal, double dy, double dx);

// Return the integral of the charge histogram above 5pc
void chargeIntegral(TH1F* charges_signal);

// Beautify charge plot
void prettyPlot(TH1F* h);

void printHelp();

const int termination_ohms = 50;

int main (int argc, char* argv[])
{

    if(argc==2 && strcmp(argv[1],"-h")==0){
        printHelp();
        return 0;
    }
    else if(argc < 6){
        cout << "Invalid command, use -h for help." << endl;
        return 0;
    }

    try{

        string channel = argv[5]; // analysis channel

        // Attribute Variables
        Attribute horiz_interval;
        Attribute vertical_gain;

        // Variables for time binning and veritcal resolution
        double dx,dy;
 
        // Variables for avg wfm histogram
        float trace_count = 0.0;
        std::vector<float> waveform_voltage;

        // ROOT histograms
        TH1F *pedestals = new TH1F("Pedestal","",5000,0.0,2);
        // This is the binning for scintillator data
        TH1F *charges_signal = new TH1F("Charge","",700,-5.0,800);

        // For reading in data
        string filename;
        H5File file;
        DataSet dataset;

        // Load in the datafiles from a txt file
        ifstream ifs (argv[1] , ifstream::in);
        ifs >> filename;


        while (ifs.good()){

            // Open HDF5 File, then HDF5 DataSet and Read in Attributes
            file.openFile(filename, H5F_ACC_RDONLY);
            ifs >> filename;
            dataset = file.openDataSet(channel);

            horiz_interval = dataset.openAttribute("horiz_interval");
            vertical_gain = dataset.openAttribute("vertical_gain");

            horiz_interval.read(PredType::NATIVE_DOUBLE, &dx);
            vertical_gain.read(PredType::NATIVE_DOUBLE, &dy);

            // Initialize the datacluster
            DataCluster * datacluster = Init_Data(&dataset);

            unsigned long window_length = datacluster->trace_length;

            /* User inputs */
            float pedestal_window = atoi(argv[3])/(dx*1e9); // pedestal window
            float signal_window = atoi(argv[4])/(dx*1e9); // signal window

            cout << "-------------------------------------------------------" << endl;
            cout << "Analyzing file:           "  << filename << endl;
            // Print the waveform parameters to screen
            cout << "Waveform parameters for:  " << channel << endl;
            cout << " * * * * * * * * * * * * * * * * * * * " << endl;
            cout << "Horizontal resolution:    " << dx << " ns" << endl;
            cout << "Verticle resolution:      " << dy << " V" << endl;
            cout << "Trace length:             " << window_length*dx*1e9 << " ns" << endl;
            cout << "Pedestal window:          " << "0 - " << pedestal_window << " ns"<< endl;
            cout << "Integration window:       " << pedestal_window << " - "
                 << (pedestal_window + signal_window) << " ns" << endl;
            cout << "-------------------------------------------------------" << endl;


            // Loop over every trace
            for (unsigned int j = 0; j < datacluster->n_traces; j++){
                cout << "Analyzing trace " << j+1 << " out of "
                     <<datacluster->n_traces<< " total traces " <<'\r';
                cout.flush();

                // Read in the data
                Read_Trace(datacluster,j);

                double pedestal = TMath::Mean (pedestal_window, datacluster->data_out)*dy; // baseline (V)
                unsigned int window_length = datacluster->trace_length; // trace length

                if(j == 0){ // resize once
                    waveform_voltage.resize(window_length);
                }

                // integrated charge (pC) 
                double charge = getCharge(pedestal_window, signal_window, datacluster, pedestal, dy, dx);

                // build waveform for debugging and avg wvm array
                for(unsigned int i = 0; i < window_length; i++){
                    waveform_voltage.at(i) += getVoltage(datacluster, pedestal, dy, i);
                }
                
                charges_signal->Fill(charge);
                pedestals->Fill(pedestal);

                trace_count++;
            }

            file.close(); // close hdf5 file
        }

        ifs.close(); // close file stream

        // Build the average waveform
        TH1F* avg_wfm = new TH1F("avg_wfvm","",waveform_voltage.size(),0,waveform_voltage.size()*dx*1e9);
        for(unsigned int i = 0; i < waveform_voltage.size(); i++){
            avg_wfm->SetBinContent(i, waveform_voltage[i]/trace_count);
        }
         
        chargeIntegral(charges_signal);
        prettyPlot(charges_signal);

        // Output Histograms to File
        if (argc > 2){
            TFile f(argv[2],"new");
            pedestals->Write();
            charges_signal->Write();
            avg_wfm->Write();
        }

        // clean-up
        delete pedestals;
        delete charges_signal;
        delete avg_wfm;

    } // end of try block

    // catch failure caused by the H5File operations
    catch(FileIException error){
        error.printError();
        return -1;
    }

    // catch failure caused by the DataSet operations
    catch(DataSetIException error){
        error.printError();
        return -1;
    }

    // catch failure caused by the DataSpace operations
    catch(DataSpaceIException error){
        error.printError();
        return -1;
    }

    // catch failure caused by the DataSpace operations
    catch(DataTypeIException error){
        error.printError();
        return -1;
    }

    return 0; // successfully terminated
}

/* Datacluster method */
DataCluster * Init_Data(DataSet *dataset){

    DataCluster * datacluster = new DataCluster[1];

    /* Get dataspace of the dataset. */
    datacluster->dataset = dataset;
    datacluster->dataspace = datacluster->dataset->getSpace();

    /* Get the dimension size of each dimension
       in the dataspace and display them. */
    hsize_t dims_out[2];
    datacluster->dataspace.getSimpleExtentDims(dims_out, NULL);
    datacluster->trace_length = (unsigned long)(dims_out[1]);
    datacluster->n_traces = (unsigned long)(dims_out[0]);

    // Data Buffer
    datacluster->data_out = new char[datacluster->trace_length]; // Data is size char
    for (unsigned long i = 0; i < datacluster->trace_length; i++) datacluster->data_out[i]= 0;

    /* Define hyperslab in the dataset. */
    datacluster->offset[0] = 0;
    datacluster->offset[1] = 0;
    datacluster->count[0] = 1;
    datacluster->count[1] = datacluster->trace_length;
    datacluster->dataspace.selectHyperslab( H5S_SELECT_SET, datacluster->count, datacluster->offset );

    /* Define the memory dataspace. */
    hsize_t dimsm[2]; /* memory space dimensions */
    dimsm[0] = dims_out[0];
    dimsm[1] = dims_out[1];
    datacluster->memspace = DataSpace( RANK_OUT, dimsm );

    /* Define memory hyperslab. */
    datacluster->offset_out[0] = 0;
    datacluster->offset_out[1] = 0;
    datacluster->count_out[0] = 1;
    datacluster->count_out[1] = datacluster->trace_length;
    datacluster->memspace.selectHyperslab( H5S_SELECT_SET, datacluster->count_out, datacluster->offset_out );

    return datacluster;
}

/* Method: Read_Trace(DataCluster *datacluster, unsigned long trace_index)
 * Updates a DataCluster datacluster so that its buffer contains trace number trace_index */
int Read_Trace(DataCluster *datacluster, unsigned long trace_index){
    datacluster->offset[0]= (hsize_t)trace_index;
    datacluster->dataspace.selectHyperslab( H5S_SELECT_SET, datacluster->count, datacluster->offset );
    datacluster->memspace.selectHyperslab( H5S_SELECT_SET, datacluster->count_out, datacluster->offset_out );
    datacluster->dataset->read( datacluster->data_out, PredType::NATIVE_CHAR, datacluster->memspace, datacluster->dataspace );
    return 0;
}

// Returns voltage of the sample
double getVoltage(DataCluster *datacluster, double pedestal, double dy, int i){

    float voltage = ((float)datacluster->data_out[i]*dy-pedestal);
    return voltage;
}

// Returns charge (pC) integrated over the signal window
double getCharge(float pedestal_window, float signal_window, DataCluster *datacluster, double pedestal, double dy, double dx){

    double charge = 0.0;
    // Charge integration starts at the end of the pedestal window
    for(int i = pedestal_window*dx*1e9; i < (pedestal_window + signal_window)*dx*1e9; i++){
        float voltage = ((float)datacluster->data_out[i]*dy-pedestal);
        charge+=(voltage*((-1000.0*dx*1e9)/termination_ohms)); // in pC
    }
    return charge;
}

// Charge-weighted integral of the charge histogram
// Compare against LAB+PPO standard
void chargeIntegral(TH1F* charges_signal){

    bool degassed = false;

    // Degassed light yield
    const float lab_ppo_ly = 1.99509e+07; 
    //const float lab_ppo_ly = 1.83045e+07;

    // Integration range
    TAxis *axis = charges_signal->GetXaxis();
    int bmin = axis->FindBin(5.0);
    int bmax = axis->FindBin(800.0);
    double x=0.0,y=0.0,err=0.0,integral=0.0,terror=0.0;
    double endpoint = 0.0;
    double bin_content_end = 10.0;

    // Charge-weighted integral of charge histogram
    for(int i = bmin; i < bmax; i++){
        x = axis->GetBinCenter(i);
        y = charges_signal->GetBinContent(i);
        err = charges_signal->GetBinError(i);
        if(y >= bin_content_end){
          endpoint = x;
        }
        integral += x*y;
        terror += x*err; 
    }

    // Print the integral of the charge distribtuion above some noise threshold
    cout << endl;
    cout << "-------------------------------------------------------" << endl;
    if(degassed){
      cout << "Comparing to degassed scintillator light yield." << endl;
    }
    else{
      cout << "Comparing to gassed scintillator light yield." << endl;
    }
    cout << "Weighted integral above " << 5.0 << "pC is " << integral << endl;
    cout << "End point (last bin with content > " << bin_content_end << ") is " << endpoint << " pC " << endl;
    cout << "Relative light yield is: " << integral/lab_ppo_ly << " +/- " << terror/lab_ppo_ly << endl;
    cout << "-------------------------------------------------------" << endl;
 
    TCanvas *c = new TCanvas;
    charges_signal->Draw();
    c->Update();
}

// Beautify charge plot
void prettyPlot(TH1F* h){

    TAxis *xaxis = h->GetXaxis();
    TAxis *yaxis = h->GetYaxis();
    xaxis->SetTitleFont(132);
    xaxis->SetLabelFont(132);
    xaxis->SetTitle("Charge (pC)");
    yaxis->SetTitleFont(132);
    yaxis->SetLabelFont(132);
    yaxis->SetTitle("Counts");
    h->SetLineColor(kBlack);
}

void printHelp(){
    cout << "Command lines arguments (all are required):" << endl;
    cout << "[1] Text file listing .hdf5 files for analysis." << endl;
    cout << "[2] Output root file." << endl;
    cout << "[3] Pedestal window length, in samples." << endl;
    cout << "[4] Integration winodw length, in samples." << endl;
    cout << "[5] Name of channel to analyze." << endl;
}
